sentinel-core/statistic/base/leap_array.h

```diff
@@ -77,23 +79,21 @@ WindowWrapSharedPtr<T> LeapArray<T>::CurrentWindow(int64_t time_millis) {
   int64_t bucket_start = CalculateWindowStart(time_millis);
 
   while (true) {
+    mtx_.Lock();
     WindowWrapSharedPtr<T> old = array_[idx];
+    mtx_.Unlock();
     if (old == nullptr) {
-      std::unique_lock<std::mutex> lck(mtx_, std::defer_lock);
-      if (lck.try_lock() && array_[idx] == nullptr) {
-        WindowWrapSharedPtr<T> bucket = std::make_shared<WindowWrap<T>>(
-            bucket_length_ms_, bucket_start, NewEmptyBucket(time_millis));
-        array_[idx] = bucket;
-        return bucket;
-      }
+      WindowWrapSharedPtr<T> bucket = std::make_shared<WindowWrap<T>>(
+          bucket_length_ms_, bucket_start, NewEmptyBucket(time_millis));
+      mtx_.Lock();
```

Comment:
Why can't I just wait directly on the lock, but do a spinlock?

Reply:
We need to guarantee that only one thread can get the lock and do the update, while other threads should spin until the update has been finished (rather than wait for the lock then do the duplicate update).

Reply:
Lock can also guarantee that only one thread is in the critical section at the same time.

Reply:
Any progress? :)